### PR TITLE
variable: add document about lazy cursor fetch and its variable

### PR DESCRIPTION
**Pull request by YangKeao**
### What is changed, added or deleted? (Required)

Add a new variable `tidb_enable_lazy_cursor_fetch`, and introduce the effect and limitation of it.

### Which TiDB version(s) do your changes apply to? (Required)

**Tips for choosing the affected version(s):**

By default, **CHOOSE MASTER ONLY** so your changes will be applied to the next TiDB major or minor releases. If your PR involves a product feature behavior change or a compatibility change, **CHOOSE THE AFFECTED RELEASE BRANCH(ES) AND MASTER**.

For details, see [tips for choosing the affected versions (in Chinese)](https://github.com/pingcap/docs-cn/blob/master/CONTRIBUTING.md#版本选择指南).

- [x] master (the latest development version)
- [x] v8.3 (TiDB 8.3 versions)
- [ ] v8.2 (TiDB 8.2 versions)
- [ ] v8.1 (TiDB 8.1 versions)
- [ ] v8.0 (TiDB 8.0 versions)
- [ ] v7.5 (TiDB 7.5 versions)
- [ ] v7.1 (TiDB 7.1 versions)
- [ ] v6.5 (TiDB 6.5 versions)
- [ ] v6.1 (TiDB 6.1 versions)
- [ ] v5.4 (TiDB 5.4 versions)
- [ ] v5.3 (TiDB 5.3 versions)
- [ ] v5.2 (TiDB 5.2 versions)

### What is the related PR or file link(s)?

- This PR is translated from: https://github.com/pingcap/docs/pull/18329
- Other reference link(s):

### Do your changes match any of the following descriptions?

- [ ] Delete files
- [ ] Change aliases
- [ ] Need modification after applied to another branch 
- [ ] Might cause conflicts after applied to another branch

### CODE REVIEW

File: system-variables.md [hunk 1]
```diff
@@ -2441,6 +2441,24 @@ Query OK, 0 rows affected (0.09 sec)
...
+
+- 作用域：GLOBAL
+- 是否持久化到集群：是
+- 是否受 Hint [SET_VAR](/optimizer-hints.md#set_varvar_namevar_value) 控制：否
+- 类型：布尔型
+- 默认值：`OFF`
+- 这个变量用于控制 [Cursor Fetch](/develop/dev-guide-connection-parameters.md#使用-streamingresult-流式获取执行结果) 功能的行为。在变量为 `OFF` 时，开启 Cursor Fetch 的情况下，TiDB 将在语句开始执行时将所有数据读取完成并保存在 TiDB 内存中，并在后续客户端读取的过程中依据客户端指定 `FetchSize` 返回给客户端，如果结果集过大，可能触发落盘临时将结果写入硬盘。在变量为 `ON` 时，开启 Cursor Fetch 的情况下，TiDB 将不会一次把所有数据读取到 TiDB 节点，而是会随着客户端的读取不断将数据读到 TiDB 节点。
+- 该功能存在如下限制：
+    - 1. 不支持处于显式事务中的语句。
+    - 2. 仅支持 `TableReader`、`IndexReader`、`IndexLookUp`、`Projection`、`Selection` 这些执行计划。
```
Comment (lcwangchao) [inline system-variables.md:2458]:
```suggestion
    - 2. 当前仅支持包含且仅包含 `TableReader`、`IndexReader`、`IndexLookUp`、`Projection`、`Selection` 算子的执行计划。
```

File: system-variables.md [hunk 2]
```diff
@@ -1926,6 +1926,24 @@ mysql> SELECT job_info FROM mysql.analyze_jobs ORDER BY end_time DESC LIMIT 1;
 - 可选值：`OFF`，`ON`
 - 这个变量用于控制是否支持对分区表创建 `Global index`。`Global index` 当前正处于开发阶段，**不推荐修改该变量值**。
 
+### `tidb_enable_lazy_cursor_fetch` <span class="version-mark">从 v8.3.0 版本开始引入</span>
+
+> **警告：**
+>
+> 该变量控制的功能为实验特性，不建议在生产环境中使用。
```
Comment (lilin90) [inline system-variables.md:1933]: 如无需要特别提及的点，建议采用通用的实验特性警告描述。
```suggestion
> 该变量控制的功能为实验特性，不建议在生产环境中使用。该功能可能会在未事先通知的情况下发生变化或删除。如果发现 bug，请在 GitHub 上提 [issue](https://github.com/pingcap/tidb/issues) 反馈。
```

File: system-variables.md [hunk 3]
```diff
@@ -1926,6 +1926,24 @@ mysql> SELECT job_info FROM mysql.analyze_jobs ORDER BY end_time DESC LIMIT 1;
...
+
+- 作用域：GLOBAL
+- 是否持久化到集群：是
+- 是否受 Hint [SET_VAR](/optimizer-hints.md#set_varvar_namevar_value) 控制：否
+- 类型：布尔型
+- 默认值：`OFF`
+- 这个变量用于控制 [Cursor Fetch](/develop/dev-guide-connection-parameters.md#使用-streamingresult-流式获取执行结果) 功能的行为。在变量为 `OFF` 时，开启 Cursor Fetch 的情况下，TiDB 将在语句开始执行时将所有数据读取完成并保存在 TiDB 内存中，并在后续客户端读取的过程中依据客户端指定 `FetchSize` 返回给客户端，如果结果集过大，可能触发落盘临时将结果写入硬盘。在变量为 `ON` 时，开启 Cursor Fetch 的情况下，TiDB 将不会一次把所有数据读取到 TiDB 节点，而是会随着客户端的读取不断将数据读到 TiDB 节点。
```
Comment (lilin90) [inline system-variables.md:1940]:
```suggestion
- 这个变量用于控制 [Cursor Fetch](/develop/dev-guide-connection-parameters.md#使用-streamingresult-流式获取执行结果) 功能的行为。
    - 当开启 Cursor Fetch 且该变量设置为 `OFF` 时，TiDB 会在语句开始执行时将所有数据读取完成并保存至 TiDB 内存，在后续客户端读取的过程中会依据客户端指定的 `FetchSize` 返回给客户端。如果结果集过大，可能会触发落盘临时将结果写入硬盘。
    - 当开启 Cursor Fetch 且该变量设置为 `ON` 时，TiDB 不会一次把所有数据读取到 TiDB 节点，而是会随着客户端的读取不断将数据读到 TiDB 节点。
```

File: system-variables.md [hunk 4]
```diff
@@ -1926,6 +1926,24 @@ mysql> SELECT job_info FROM mysql.analyze_jobs ORDER BY end_time DESC LIMIT 1;
...
+- 该功能存在如下限制：
+    - 1. 不支持处于显式事务中的语句。
+    - 2. 当前仅支持包含且仅包含 `TableReader`、`IndexReader`、`IndexLookUp`、`Projection`、`Selection` 算子的执行计划。
+    - 3. 对于使用 Lazy Cursor Fetch 的语句，执行信息将不会出现在 [Statements Summary](/statement-summary-tables.md) 和 [Slow Log](/identify-slow-queries.md) 中。
```
Comment (lilin90) [inline system-variables.md:1944]:
```suggestion
- 该功能存在以下限制：
    - 不支持处于显式事务中的语句。
    - 当前仅支持包含且仅包含 `TableReader`、`IndexReader`、`IndexLookUp`、`Projection`、`Selection` 算子的执行计划。
    - 对于使用 Lazy Cursor Fetch 的语句，执行信息将不会出现在 [Statements Summary](/statement-summary-tables.md) 和[慢查询日志](/identify-slow-queries.md)中。
```

File: system-variables.md [hunk 5]
```diff
@@ -1926,6 +1926,24 @@ mysql> SELECT job_info FROM mysql.analyze_jobs ORDER BY end_time DESC LIMIT 1;
...
+- 对于暂不支持的场景，行为将会与变量为 `OFF` 时保持一致。
```
Comment (lilin90) [inline system-variables.md:1945]:
```suggestion
- 对于暂不支持的场景，行为与变量设置为 `OFF` 时一致。
```

File: system-variables.md [hunk 6]
```diff
@@ -1926,6 +1926,26 @@ mysql> SELECT job_info FROM mysql.analyze_jobs ORDER BY end_time DESC LIMIT 1;
...
+
+- 作用域：GLOBAL
+- 是否持久化到集群：是
+- 是否受 Hint [SET_VAR](/optimizer-hints.md#set_varvar_namevar_value) 控制：否
+- 类型：布尔型
+- 默认值：`OFF`
+- 这个变量用于控制 [Cursor Fetch](/develop/dev-guide-connection-parameters.md#使用-streamingresult-流式获取执行结果) 功能的行为。
+    - 当开启 Cursor Fetch 且该变量设置为 `OFF` 时，TiDB 会在语句开始执行时将所有数据读取完成并保存至 TiDB 内存，在后续客户端读取的过程中会依据客户端指定的 `FetchSize` 返回给客户端。如果结果集过大，可能会触发落盘临时将结果写入硬盘。
+    - 当开启 Cursor Fetch 且该变量设置为 `ON` 时，TiDB 不会一次把所有数据读取到 TiDB 节点，而是会随着客户端的读取不断将数据读到 TiDB 节点。
+- 该功能存在以下限制：
```
Comment (lilin90) [inline system-variables.md:1943]:
```suggestion
- 该变量控制的功能存在以下限制：
```

File: system-variables.md [hunk 7]
```diff
@@ -1926,6 +1926,26 @@ mysql> SELECT job_info FROM mysql.analyze_jobs ORDER BY end_time DESC LIMIT 1;
...
+    - 不支持处于显式事务中的语句。
+    - 当前仅支持包含且仅包含 `TableReader`、`IndexReader`、`IndexLookUp`、`Projection`、`Selection` 算子的执行计划。
+    - 对于使用 Lazy Cursor Fetch 的语句，执行信息将不会出现在 [Statements Summary](/statement-summary-tables.md) 和 [慢查询日志](/identify-slow-queries.md) 中。
+- 对于暂不支持的场景，行为与变量设置为 `OFF` 时一致。
```
Comment (lilin90) [inline system-variables.md:1947]:
```suggestion
- 对于暂不支持的场景，其行为与将变量设置为 `OFF` 时一致。
```

File: system-variables.md [hunk 8]
```diff
@@ -1926,6 +1926,26 @@ mysql> SELECT job_info FROM mysql.analyze_jobs ORDER BY end_time DESC LIMIT 1;
...
+- 是否受 Hint [SET_VAR](/optimizer-hints.md#set_varvar_namevar_value) 控制：否
+- 类型：布尔型
+- 默认值：`OFF`
+- 这个变量用于控制 [Cursor Fetch](/develop/dev-guide-connection-parameters.md#使用-streamingresult-流式获取执行结果) 功能的行为。
+    - 当开启 Cursor Fetch 且该变量设置为 `OFF` 时，TiDB 会在语句开始执行时将所有数据读取完成并保存至 TiDB 内存，在后续客户端读取的过程中会依据客户端指定的 `FetchSize` 返回给客户端。如果结果集过大，可能会触发落盘临时将结果写入硬盘。
+    - 当开启 Cursor Fetch 且该变量设置为 `ON` 时，TiDB 不会一次把所有数据读取到 TiDB 节点，而是会随着客户端的读取不断将数据读到 TiDB 节点。
+- 该功能存在以下限制：
+    - 不支持处于显式事务中的语句。
+    - 当前仅支持包含且仅包含 `TableReader`、`IndexReader`、`IndexLookUp`、`Projection`、`Selection` 算子的执行计划。
+    - 对于使用 Lazy Cursor Fetch 的语句，执行信息将不会出现在 [Statements Summary](/statement-summary-tables.md) 和 [慢查询日志](/identify-slow-queries.md) 中。
```
Comment (lilin90) [inline system-variables.md:1946]:
```suggestion
    - 对于使用 Lazy Cursor Fetch 的语句，执行信息将不会出现在 [statements summary](/statement-summary-tables.md) 和[慢查询日志](/identify-slow-queries.md)中。
```

**Review by lilin90 — APPROVED** [review]
/lgtm

File: system-variables.md [hunk 9]
```diff
@@ -1926,6 +1926,26 @@ mysql> SELECT job_info FROM mysql.analyze_jobs ORDER BY end_time DESC LIMIT 1;
...
+
+> **警告：**
+>
+> 该变量控制的功能为实验特性，不建议在生产环境中使用。该功能可能会在未事先通知的情况下发生变化或删除。如果发现 bug，请在 GitHub 上提 [issue](https://github.com/pingcap/tidb/issues) 反馈。
+
+- 作用域：GLOBAL
+- 是否持久化到集群：是
+- 是否受 Hint [SET_VAR](/optimizer-hints.md#set_varvar_namevar_value) 控制：否
+- 类型：布尔型
+- 默认值：`OFF`
```
Comment (hfxsd) [inline system-variables.md:1939]:
```suggestion
- 默认值：`OFF`
- 可选值：`OFF`，`ON`
```

**Review by hfxsd — APPROVED** [review]

**Comment by songrijie** [comment]
please review this page and see if we shall add some descriptions for this change. 
https://docs.pingcap.com/tidb/stable/dev-guide-connection-parameters#use-streamingresult-to-get-the-execution-result

File: develop/dev-guide-connection-parameters.md [hunk 10]
```diff
@@ -138,6 +138,10 @@ Java 应用尽管可以选择在不同的框架中封装，但在最底层一般
 
 TiDB 中同时支持两种方式，但更推荐使用第一种将 **FetchSize** 设置为 `Integer.MIN_VALUE` 的方式，比第二种功能实现更简单且执行效率更高。
 
+对于第二种方法，TiDB 会先将所有数据加载到 TiDB 节点上，然后根据 `FetchSize` 依次返回给客户端。因此，通常会比第一种方法使用更多的内存，如果 [`tidb_enable_tmp_storage_on_oom`](/system-variables.md#tidb_enable_tmp_storage_on_oom) 为 `ON` 时，可能会将数据存储到临时磁盘中。
+
+如果系统变量 [`tidb_enable_lazy_cursor_fetch`](/system-variables.md#tidb_enable_lazy_cursor_fetch) 为 `ON`，TiDB 将尝试仅在客户端请求数据时读取部分数据，以使用更少的内存。有关详细信息和限制，请阅读变量 [`tidb_enable_lazy_cursor_fetch`](/system-variables.md#tidb_enable_lazy_cursor_fetch) 的描述。
```
Comment (lilin90) [inline develop/dev-guide-connection-parameters.md:143]:
```suggestion
如果系统变量 [`tidb_enable_lazy_cursor_fetch`](/system-variables.md#tidb_enable_lazy_cursor_fetch-从-v830-版本开始引入) 为 `ON`，TiDB 将尝试仅在客户端请求数据时读取部分数据，以使用更少的内存。有关详细信息和限制，请阅读变量 [`tidb_enable_lazy_cursor_fetch`](/system-variables.md#tidb_enable_lazy_cursor_fetch) 的描述。
```

File: develop/dev-guide-connection-parameters.md [hunk 11]
```diff
@@ -133,10 +133,14 @@ Java 应用尽管可以选择在不同的框架中封装，但在最底层一般
 
 在 JDBC 中通常有以下两种处理方式：
 
-- 设置 [**FetchSize** 为 `Integer.MIN_VALUE`](https://dev.mysql.com/doc/connector-j/en/connector-j-reference-implementation-notes.html#ResultSet) 让客户端不缓存，客户端通过 StreamingResult 的方式从网络连接上流式读取执行结果。
-- 使用 Cursor Fetch，首先需[设置 **FetchSize**](http://makejavafaster.blogspot.com/2015/06/jdbc-fetch-size-performance.html) 为正整数，且在 JDBC URL 中配置 `useCursorFetch = true`。
+- 方式一：设置 [**FetchSize** 为 `Integer.MIN_VALUE`](https://dev.mysql.com/doc/connector-j/en/connector-j-reference-implementation-notes.html#ResultSet) 让客户端不缓存，客户端通过 StreamingResult 的方式从网络连接上流式读取执行结果。
+- 方式二：使用 Cursor Fetch，首先需[设置 **FetchSize**](http://makejavafaster.blogspot.com/2015/06/jdbc-fetch-size-performance.html) 为正整数，且在 JDBC URL 中配置 `useCursorFetch = true`。
 
-TiDB 中同时支持两种方式，但更推荐使用第一种将 **FetchSize** 设置为 `Integer.MIN_VALUE` 的方式，比第二种功能实现更简单且执行效率更高。
+TiDB 同时支持以上两种方式，但更推荐使用第一种将 **FetchSize** 设置为 `Integer.MIN_VALUE` 的方式，比第二种功能实现更简单且执行效率更高。
```
Comment (lilin90) [inline develop/dev-guide-connection-parameters.md:139]:
```suggestion
TiDB 同时支持以上两种方式，但更推荐使用第一种将 `FetchSize` 设置为 `Integer.MIN_VALUE` 的方式，比第二种功能实现更简单且执行效率更高。
```

File: best-practices/java-app-best-practices.md [hunk 12]
```diff
@@ -56,15 +56,19 @@ Java 应用尽管可以选择在不同的框架中封装，但在最底层一般
...
 
     如果需要在 `resultset` 读取完成或 close 前进行查询避免上述报错，可在 URL 中添加配置参数 `clobberStreamingResults=true`，这样会自动 close `resultset`，但之前流式查询未被读取的结果集会丢失。
 
-- 使用 Cursor Fetch，首先需[设置 `FetchSize`](http://makejavafaster.blogspot.com/2015/06/jdbc-fetch-size-performance.html) 为正整数，且在 JDBC URL 中配置 `useCursorFetch = true`。
+- 方式二：使用 Cursor Fetch，首先需[设置 `FetchSize`](http://makejavafaster.blogspot.com/2015/06/jdbc-fetch-size-performance.html) 为正整数，且在 JDBC URL 中配置 `useCursorFetch = true`。
 
-TiDB 中同时支持两种方式，但更推荐使用第一种将 `FetchSize` 设置为 `Integer.MIN_VALUE` 的方式，比第二种功能实现更简单且执行效率更高。
+TiDB 同时支持以上两种方式，但更推荐使用第一种将 `FetchSize` 设置为 `Integer.MIN_VALUE` 的方式，比第二种功能实现更简单且执行效率更高。
+
+对于第二种方式，TiDB 会先将所有数据加载到 TiDB 节点上，然后根据 `FetchSize` 依次返回给客户端。因此，通常会比第一种方式使用更多内存。如果将 [`tidb_enable_tmp_storage_on_oom`](/system-variables.md#tidb_enable_tmp_storage_on_oom) 设置为 `ON`，可能会将数据存储到临时磁盘中。
```
Comment (hfxsd) [inline best-practices/java-app-best-practices.md:69]: 最后一句话是否要和[系统变量中的描述](https://github.com/pingcap/docs-cn/pull/18073/files#diff-ec2cbb115cc58fe3afe2862eb5148732d323cbf7ae35dea97bd381a1f8d0e8a0R1942)保持一致？

可能会触发落盘临时将结果写入硬盘。

**Review by lcwangchao — APPROVED** [review]

**Pull request merged**